### DISCUSSION TITLE
symmetry breaking when building a model with python

### DISCUSSION THREAD

**Discussion by chentianhao-TUD** [google/or-tools#3295] [CP-SAT questions]
Hi,

i am new to ortools and constraint programming. I am using Python API to build a VRP model. The solver is CP-SAT. I wonder if i can speed up the solver by breaking the symmetries. How can i add some symmetries breaking constraints to the model? 

Thanks!

**Comment by lperron** [comment]
why not use the routing library ?

**Reply by chentianhao-TUD** [reply]
Hi Laurent,
thanks a lot for answering the question! I use cp sat because I want to solve vrp problem with constraint programming. And I think routing library has too much functionalities for beginner to start with. My project is to translate a gurobi MIP model into a CP model. The model has 2 types of vehicles. One is Transporter and the other is Robots. The Transporter carries robots from one node to an other. It dosen´t matter which transporter carry which robot and visit which node first. So I think there should be a lot of symmertries in the solution space. If I can eliminate them in the model rather then leave them to the solver, I should be able to solve the problem in a very short time.

**Comment by lperron** [comment]
Have you tried without symmetry breaking ?
Have you enabled parallelism ?

Can you improve the model without touching the symmetry part ?
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le jeu. 12 mai 2022 à 20:24, chentianhao-TUD ***@***.***> a
écrit :

**Reply by chentianhao-TUD** [reply]
yes I have tried without symmetry breaking. I didn't make any setting about parallelism. I am trying to improve the model without touching symmetry. I just don't know what else i can do. I will try to explain my idea quickly, maybe you can give me some advise. The objective of the model is to minimize the maximum of the time that the vehicle returning to depot. I have the boolean variable x[k, i, j] to indicate whether transporter k is visiting j from i and w[r, i, j] for the robots. The robots needs transporter to carry them and robots will serve the customers. For each vehicle I create a list task[k] or task[r] to store the OptionalIntervalVar which is performed if x[k, i, j] or w[r, i, j] is true. Then i can use the NoOverlap constraint on these lists. For each vehicle i also create a list to store the arcs like [i, j, x[k, i, j]]. Then I can use AddCircuit on them. Overall i wan to make sure the vehicle starts from depots and then return to depots. All customer (10 nodes) got visited once and if a vehicle visits a node also leaves it. But the model is obviously not good since the run time is pretty long. Are the symmetries being handled with during the presolve? How should I improve the model?

**Comment by lperron** [comment]
Set the parameter num_search_workers to 8 or more. This should be more
beneficial than symmetry breaking with a smaller effort.

Le ven. 13 mai 2022, 20:54, chentianhao-TUD ***@***.***> a
écrit :
